### PR TITLE
Improve payload type for resolver non nullable error

### PR DESCRIPTION
**Pull request by do4gr**
* improve the error message if a resolver has a non-nullable return type but returns null
* bubble up errors for invalid function response instead of throwing

* this enables us to log invalid responses that previously didn't appear in the log (like providing neither data nor error object for a non-nullable payload)

### CODE REVIEW

File: server/backend-shared/src/main/scala/cool/graph/shared/errors/Errors.scala [hunk 1]
```diff
@@ -20,7 +20,7 @@ object CommonErrors {
   case class QueriesNotAllowedForProject(projectId: String) extends UserFacingError(s"Queries are not allowed for the project with id '$projectId'", 1002)
 
   case class MutationsNotAllowedForProject(projectId: String)
-      extends UserFacingError(s"The project '$projectId' is currently in read-only mode. Please try again in a few minutes", 1003)
+    extends UserFacingError(s"The project '$projectId' is currently in read-only mode. Please try again in a few minutes", 1003)
```
Comment (dpetrick) [inline server/backend-shared/src/main/scala/cool/graph/shared/errors/Errors.scala]: Are you sure your scalafmt is set up correctly?

File: server/client-shared/src/main/scala/cool/graph/client/requestPipeline/FunctionExecutor.scala [hunk 2]
```diff
@@ -121,7 +129,7 @@ class FunctionExecutor(implicit val inj: Injector) extends Injectable {
           logsPublisher.publish(renderLogPayload("FAILURE", Map("error" -> s"Couldn't parse response: $badBody. Error message: $parseError")))
 
         case Success(Bad(FunctionReturnedBadStatus(statusCode, rawResponse))) =>
-          logsPublisher.publish(renderLogPayload("FAILURE", Map("error" -> rawResponse))) //Function returned invalid status code: $statusCode. Raw body:
+          logsPublisher.publish(renderLogPayload("FAILURE", Map("error" -> s"Function returned invalid status code: $statusCode. Raw body: $rawResponse")))
```
Comment (dpetrick) [inline server/client-shared/src/main/scala/cool/graph/client/requestPipeline/FunctionExecutor.scala]: This is reverting an intentional change, what is the reasoning for changing it back?

File: server/client-shared/src/main/scala/cool/graph/client/requestPipeline/FunctionExecutor.scala [hunk 3]
```diff
@@ -143,29 +153,58 @@ class FunctionExecutor(implicit val inj: Injector) extends Injectable {
...
+          tryParsingAsList(data) match {
+            case Good(vector: Vector[Any]) =>
+              val parsed: Vector[Or[JsObject, FunctionError]] = vector.map(parseDataToJsObject)
+              val error                                       = parsed.find(_.isBad)
+
+              error match {
+                case None           => Good(FunctionDataValue(isNull = false, parsed.collect { case Good(jsObject) => jsObject }))
+                case Some(Bad(err)) => Bad(err)
+                case _              => Bad(FunctionReturnValueParsingError(function.name))
+              }
```
Comment (dpetrick) [inline server/client-shared/src/main/scala/cool/graph/client/requestPipeline/FunctionExecutor.scala]: Style: You forgot a blank line here.

File: server/client-shared/src/main/scala/cool/graph/client/requestPipeline/FunctionExecutor.scala [hunk 4]
```diff
@@ -191,10 +230,14 @@ class FunctionExecutor(implicit val inj: Injector) extends Injectable {
               FunctionExecutionResult(Vector.empty, parsed)
           }
 
-          def getResult(data: Any): FunctionDataValue = function match {
+          def getResult(data: Any): FunctionDataValue Or FunctionError = function match {
             case f: CustomQueryFunction    => parseResolverResponse(data, f.payloadType)
```
Comment (dpetrick) [inline server/client-shared/src/main/scala/cool/graph/client/requestPipeline/FunctionExecutor.scala]: Style: If one case exceeds one line, break all cases to newlines and have a blank line in between each case.

File: server/client-shared/src/test/scala/cool/graph/private_api/finder/CachedProjectFetcherImplSpec.scala [hunk 5]
```diff
@@ -56,7 +56,7 @@ class CachedProjectFetcherImplSpec extends FlatSpec with Matchers with ScalaFutu
     projectFetcher.setAlias(firstAlias = None, secondAlias = None)
     pubSub.publish(Only("FirstOne"), "FirstOne")
 
-    Thread.sleep(2000)
+    Thread.sleep(3000)
```
Comment (dpetrick) [inline server/client-shared/src/test/scala/cool/graph/private_api/finder/CachedProjectFetcherImplSpec.scala]: Didn't you do this in a previous PR as well?

Comment (dpetrick) [reply server/client-shared/src/test/scala/cool/graph/private_api/finder/CachedProjectFetcherImplSpec.scala]: See https://github.com/graphcool/framework/pull/1096/files/d410409b82bc14133aea2e5b91e362f90e003b3c#diff-7f784c4dd4c366679489718601b496f3R61

File: server/client-shared/src/main/scala/cool/graph/client/requestPipeline/FunctionExecutor.scala (hunk 4 above)
Comment (do4gr) [reply server/client-shared/src/main/scala/cool/graph/client/requestPipeline/FunctionExecutor.scala]: done

File: server/backend-shared/src/main/scala/cool/graph/shared/errors/Errors.scala (hunk 1 above)
Comment (do4gr) [reply server/backend-shared/src/main/scala/cool/graph/shared/errors/Errors.scala]: It wasn't 👍

**Pull request merged**